File: MRCRWidget-Example-Objc/ViewController.m
//
//  Copyright (c) 2022-present MoneySwap OU (mercuryo.io). All rights reserved.
//

#import "ViewController.h"

@import MRCRWidget;

@interface ViewController () <MRCRWidgetUIDelegate>

@property (nonatomic, weak) UIViewController *widgetViewController;

@end

@implementation ViewController

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    
    [self showWidget];
}

- (void)showWidget {
    MRCRWidgetError *configurationError;
    
    MRCRWidgetConfiguration *configuration = [[MRCRWidgetConfiguration alloc]
                                              initWithWidgetId:@""
                                              environment:MRCRWidgetEnvironmentProduction
                                              params:NULL
                                              error:&configurationError];
    if (configurationError) {
        [NSException raise:@"FATAL ERROR" format:@"%@", configurationError.localizedDescription];
        return;
    }
    
    MRCRWidgetAssembly *assembly = [[MRCRWidgetAssembly alloc] initWithConfiguration:configuration
                                                                          uiDelegate:self];
    UIViewController *vc = [assembly build];
    
    [self presentViewController:vc
                       animated:YES
                     completion:NULL];
    
    self.widgetViewController = vc;
}

- (void)showAlert {
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Access Denied"
                                                                             message:@"Please grant access to your photo library"
                                                                      preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction *action = [UIAlertAction actionWithTitle:@"OK"
                                                     style:UIAlertActionStyleDefault
                                                   handler:NULL];
    [alertController addAction:action];
    
    [self.widgetViewController presentViewController:alertController
                                            animated:YES
                                          completion:NULL];
}

#pragma mark - MRCRWidgetUIDelegate

- (void)mrcrWidgetDidReceiveError:(MRCRWidgetError * _Nonnull)error {
    switch (error.errorCode) {
        case MRCRWidgetErrorCodePhotoLibraryAuthorizationFailed:
            [self showAlert];
            break;
            
        default:
            break;
    }
}

@end

File: Example-Swift/MRCRWidget/ViewController.swift
//
//  Copyright (c) 2022-present MoneySwap OU (mercuryo.io). All rights reserved.
//

import UIKit
import MRCRWidget


final class ViewController: UIViewController {
    
    
    // MARK: - Private Properties
    
    private weak var widgetViewController: UIViewController?
    
    
    // MARK: - Life Cycle
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        
        showWidget()
    }
    
    
    // MARK: - Private Methods
    
    private func showWidget() {
        do {
            let configuration = try WidgetConfiguration(widgetId: "")
            let vc = WidgetAssembly(
                configuration: configuration,
                uiDelegate: self
            )
                .build()
            
            self.present(vc, animated: true)
            self.widgetViewController = vc
        } catch {
            assertionFailure(error.localizedDescription)
        }
    }
    
    private func showAlert() {
        let alertController = UIAlertController(
            title: "Access Denied",
            message: "Please grant access to your photo library",
            preferredStyle: .alert
        )
        alertController.addAction(
            UIAlertAction(
                title: "OK",
                style: .default
            )
        )
        
        widgetViewController?.present(
            alertController,
            animated: true
        )
    }
}


// MARK: - WidgetUIDelegate

extension ViewController: WidgetUIDelegate {
    
    func widgetDidReceiveError(_ error: WidgetError) {
        switch error.errorCode {
        case .photoLibraryAuthorizationFailed:
            showAlert()
            
        default:
            break
        }
    }
}
